### DISCUSSION TITLE
No ambient temp on XTRA 4415N v2.0 firmware

### DISCUSSION THREAD

**Discussion by murfy76** [chickey/RS485-WiFi-EPEver#20] [General]
Hello, 

thanks for this great project. I have ESP8266 NodeMcu RX/TX pins connected to MAX485 module whithout RE and DE pins. Everything is fine except I cant see ambient temp value. What I can do to read ambient temperature? Temp sensor is connected to controller and Solar Station Monitor can read values. Thank you.

Peter

**Comment by chickey** [comment]
Hi Peter,   you would have to find the register and then modify the software to read that specific register and display it.

Colin

**Comment by murfy76** [comment]
Hi Colin,

Thank you. I found specific registers 0x3110 for battery_temperature here https://esphome.io/cookbook/tracer-an.html?highlight=epsolar. Register should go after loadPower, so I changed code in config.h to:
...
  // live data
  union {
    struct {
      ...
      int32_t  lP;
      int16_t  bT; 
    } l;
    uint16_t  buf[17];
  } live; 
...
also changed "#define LIVE_DATA_CNT" to 17 and in mqtt.h added:

  sprintf(buf,"%s/battery/T",myConfig.mqtt_topic);
  mqtt_publish_f( buf, live.l.bT /100.f);
...
  panelDoc[F("netI")] = batteryCurrent/100.0f;
  panelDoc[F("T")] = live.l.bT /100.f;

Any idea what's wrong here? ESP stop working after this change. My Epever controller is hardly accessible so it's not easy to flash it back to default. Thanks.

Peter

**Reply by chickey** [reply]
Hi Peter,
             I've configured the changes you've written manually and it works on my Wemos D1 mini ok.

Thanks
Colin

**Reply by murfy76** [reply]
Hi Colin,

finally success on third attempt :-) However OTA upgrade was not successful again. I had to connect Nodemcu to USB and erase flash first. Than flash new firmware and setup connection to mqtt. I'm able to read battery (ambient) and device temperature now. If someone interested changes are bellow. Thanks.

Peter

--- config.h:
+ #define LIVE_DATA_CNT   18
...
      int32_t  lP;
+    int16_t  bT;
+    int16_t  dT; 
    } l;
+   uint16_t  buf[18];
  } live;
 
--- mqtt.h:
  mqtt_publish_f( buf, live.l.lP /100.f);

+  sprintf(buf,"%s/temperature/bT",myConfig.mqtt_topic);
+  mqtt_publish_f( buf, live.l.bT /100.f);
+  sprintf(buf,"%s/temperature/dT",myConfig.mqtt_topic);
+  mqtt_publish_f( buf, live.l.dT /100.f);
...
  sprintf_P(mqtt_topic, PSTR("%s/battery"), myConfig.mqtt_topic);
  jsonLength = serializeJson(panelDoc, buf);
  mqtt_client.publish(mqtt_topic, buf, jsonLength);
  panelDoc.clear();

+  //temperature
+  panelDoc[F("bT")] = live.l.bT /100.f;
+  panelDoc[F("dT")] = live.l.dT /100.f;

+  sprintf_P(mqtt_topic, PSTR("%s/temperature"), myConfig.mqtt_topic);
+  jsonLength = serializeJson(panelDoc, buf);
+  mqtt_client.publish(mqtt_topic, buf, jsonLength);
+  panelDoc.clear();
